### ISSUE TITLE
Should we have a "ShutdownException" class?

### ISSUE THREAD

**Issue by deanebarker** [statiqdev/Statiq#219]
Looking down the road, I wonder if we shouldn't have a special Exception for problems that are _intended_ to halt execution.  We're just throwing generic exceptions, which is fine in situations where the problem was unforeseen, but there are "deliberate exception" situations where we're throwing an exception _specifically_ to halt execution (the entire ValidateMeta module, for instance).

Two reasons:
1. It would probably be helpful in the future to identify these exceptions.  If the exception was of this type, the context would know that this was a deliberate shutting down.
2. We could attach the "offending" document. A lot of these situations are going to happen because of a problem with a single document.  It would be nice to throw the exception, and attach the document as if to say, "I am shutting this down because _this specific document_ is mis-behaving."  This would improve our ability to report and log errors.

Also, this would be helpful in a ideal situation where we could run a "test execution."  I'd like to see a framework in the future where we could do something like "test this build," which means that someone could run an entire execution then _see the errors that came back_.  Instead of stopping at the first exception, execution would just store that exception can carry on, then display all those problems, so that they could be corrected.

**Comment by daveaglick** [comment]
I agree. Perhaps not `ShutdownException` (since every exception is a "shutdown exception"), but a typed `WyamException` is probably overdue with the ability to specify document as you suggested along with having the current pipeline and module injected. Probably two ways to do this (I think I prefer the latter):
- Create a `IExecutionContext.GetException(...)` method that instantiates and sets the module, pipeline, and any other information passed in the `GetException()` method in the exception.
- Handle `WyamException` throws at the pipeline level and inject the extra information there before rethrowing.

**Comment by deanebarker** [comment]
Would you be able to throw a `WyamException` from inside a module?  If you just bubble the raw exception up to the pipeline, how would it know what document caused the problem?

**Comment by daveaglick** [comment]
You'll probably have to give the exception (or the `IExecutionContext.GetException(...)` method) the document at fault, if there is one. Since documents can be processed in parallel inside the module, and the module code is outside our control from Core, there's no way for us to know which one caused the exception. The module and pipeline we can figure out though because that execution is under our control so we can place try-catch blocks around individual module execution calls.

**Comment by daveaglick** [comment]
But yeah, the goal here would be to make a standard exception that modules would throw if there were errors. They'll just have to provide a little bit of context by way of telling us which document was related to the problem.

**Comment by deanebarker** [comment]
Why wouldn't we just do this, from inside a module's `Execute` method (assuming `input` is the document in question):

```
throw new WyamException(input);
```

**Comment by daveaglick** [comment]
That's exactly what I was thinking.

The injection of module and pipeline information will be transparent to the module, and is mainly just to provide some context in the embedded scenario where the engine is being run from a host application (we probably don't care that much in the wyam.exe console app because we have the trace messages for context).
